text/3778-proc-macro-in-same-package-as-app.md

```diff
@@ -0,0 +1,171 @@
+
+Any libraries to be linked, as specified in `build.rs` via stdout, would be linked and made available in the `macros`.
+
+The compiled macros crate would be passed into rustc with `--extern=macros=target/_profile_/deps/lib_____` when compiling the other crates. 
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+None at the moment
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+1. Use `crate::macros::*` as the import path
+
+This would require changes to rustc, when there is a simpler solution.
+
+2. Have it within a `proc_macros` directory
+
+This was the original idea; but upon further consideration it turns out to be worse than the current. The justification of it over a file was:
+> *Since it would compile first, proc macro files cannot import functions in the main code. The problem is having it side-by-side to the rest of your code makes it seem like you could just import it, when you cannot. Having it as a seperate directory makes clear of this.*
+
+While it was true that proc macro files cannot import functions in the main code, it can import other modules, making the statement's merits false.
+
+3. Eliminate the need for new proc macro files/folders entirely, have the compiler work out where the proc macros are and separate them.
+
+This would suffer from the same issue as the last alternative, plus being harder to implement.
+
+4. Introspection
+
+Harder to implement, with less payoff relative to the amount of work required. 
+
+# Prior art
+[prior-art]: #prior-art
+
+1. Zig comptime: metaprogramming code can sit directly next to application code.
+2. Declarative macros: can sit side by side as well, but is less powerful.
+3. Lisp macros: same as last two, except more powerful.
+4. `tests` directory, and `build.rs`: compiled at a different time as the main code.
+5. `Makefiles`, or other build systems: they allow for customisability for when code is built.
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+1. Should proc macro dependencies be listed under `[build-dependencies]`, or a new `[proc-macro-dependencies]` section?
+2. What case should `procmacros.rs` be? No spaces, kebab case, or snake? Having no spaces would be the most agnostic solution
+2. ~~Should we import like `crate::proc_macro::file::macro`, or via a new keyword, like `crate_macros::file::macro`? The latter would avoid name collisions, but might be more confusing.~~
+
+# Future possibilities
+[future-possibilities]: #future-possibilities
+
+1. As described in the [motivation] section, this proposal is aimed to make the process of creating proc macros easier. So a natural extension of this is to remove the need of third-party libraries like syn and proc-macro2. There is already an effort to implement quote, so they might be a possibility.
+
+2. This might enable for some sort of `$crate` metavariable.
```

Comment:
Been thinking about this more and I'm unsure.  If we tell the proc macro at compilation time, Cargo would need to know a unique identifier for the re-exporting crate.  However, I'm not sure how easily we can determine that unique identifier when building a dependent.

If we could somehow bind `$crate` later in the process, that might work but Cargo won't know what all to apply that binding to.

If we can't do this, how much would this affect this proposal?

Reply:
I don't see how this is related to this RFC at all. Every dependent is allowed to rename library names as necessary, so it is no easier to support `$crate` in in-package proc-macros than traditional ones.

Reply:
What is improved with this RFC is that cargo now knows what API a derive is generating code against.  We could potentially find a way to communicate this to rustc so it could provide `$crate` like functionality.

See also the end of https://github.com/rust-lang/rfcs/pull/3826#discussion_r2130102665